File: ref/RobStride_Control/python/src/position_control.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
RobStride MIT 模式位置控制 (最简可靠版)
模式: Mode 0 (MIT Mode)
通信: 循环调用 write_operation_frame

用法: python3 position_control_mit.py <motor_id>
"""

import sys
import os
import time
import math
import struct
import threading
import signal
from typing import Optional

# 尝试导入 SDK
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
try:
    from robstride_dynamics import RobstrideBus, Motor, ParameterType, CommunicationType
except ImportError:
    # 假设当前目录结构
    try:
        from bus import RobstrideBus, Motor
        from protocol import ParameterType, CommunicationType
    except ImportError as e:
        print(f"❌ 无法导入 SDK: {e}")
        sys.exit(1)

class PositionControllerMIT:
    def __init__(self, motor_id: int, channel='can0'):
        self.motor_id = motor_id
        self.motor_name = f"motor_{motor_id}"
        self.channel = channel
        
        self.bus: Optional[RobstrideBus] = None
        self.lock = threading.Lock()  # 互斥锁，防止Socket冲突
        
        self.running = True
        self.connected = False
        self.target_position = 0.0  # 目标位置 (rad)
        
        # 默认参数 (MIT 模式)
        self.kp = 30.0  # 刚度 (Nm/rad)
        self.kd = 0.5   # 阻尼 (Nm/rad/s)

    def _signal_handler(self, signum, frame):
        self.stop_and_exit()

    def _set_mode_raw(self, mode: int):
        """
        使用原始 transmit 发送模式切换指令，不等待回包 (避免connect超时)
        """
        print(f"⚙️ 切换模式 (Mode {mode}) - [Raw Transmit]")
        device_id = self.bus.motors[self.motor_name].id
        param_id, param_dtype, _ = ParameterType.MODE

        # MODE 是 int8
        value_buffer = struct.pack("<bBH", mode, 0, 0)
        data = struct.pack("<HH", param_id, 0x00) + value_buffer

        self.bus.transmit(CommunicationType.WRITE_PARAMETER, self.bus.host_id, device_id, data)
        time.sleep(0.1) # 等待电机切换模式
        print(f"✅ 模式切换指令已发送")

    def connect(self):
        print(f"🔍 正在连接 CAN 通道 {self.channel}...")
        
        # 定义电机
        motors = {
            self.motor_name: Motor(id=self.motor_id, model="rs-03") # 模型型号请根据实际修改
        }
        
        # 简单的校准参数
        calibration = {
            self.motor_name: {"direction": 1, "homing_offset": 0.0}
        }

        try:
            self.bus = RobstrideBus(self.channel, motors, calibration)
            self.bus.connect(handshake=True)
            
            with self.lock:
                # 激活电机
                print(f"⚡ 激活电机 ID: {self.motor_id} ...")
                self.bus.enable(self.motor_name)
                time.sleep(0.5)

                # *********************
                # *** 核心逻辑 ***
                # *********************
                # 1. 切换到 MIT 模式 (Mode 0)
                self._set_mode_raw(0)
                
                # 2. 设置一个已知的、安全的初始目标
                print("🏠 设置初始目标为 0.0 ...")
                self.target_position = 0.0 # 设为 0 弧度
                
                # 3. 发送第一帧 MIT 指令来保持位置
                self.bus.write_operation_frame(
                    self.motor_name,
                    self.target_position,
                    self.kp,
                    self.kd,
                    0.0, # velocity_ff
                    0.0  # torque_ff
                )
                print(f"🏠 初始目标已设为: 0.0°")
            
            self.connected = True
            
            # 启动后台控制线程
            self.control_thread = threading.Thread(target=self.loop, daemon=True)
            self.control_thread.start()
            
            print("✅ 初始化完成 (Mode 0)！")
            return True
            
        except Exception as e:
            print(f"❌ 连接失败: {e}")
            self.connected = False
            return False

    def loop(self):
        """控制线程：持续发送 MIT 帧以保持位置"""
        print("🔄 控制循环已启动 (Mode 0 @ 50Hz)")
        
        while self.running and self.connected:
            try:
                with self.lock:
                    # 1. 发送 MIT 帧 (只发)
                    self.bus.write_operation_frame(
                        self.motor_name,
                        self.target_position,
                        self.kp,
                        self.kd,
                        0.0, # velocity_ff
                        0.0  # torque_ff
                    )
                    
                    # 2. 读取状态帧 (只收)
                    # 这一步至关重要，用于清空 CAN 接收缓冲区，防止溢出
                    # 我们可以忽略返回值，因为我们只关心 "清空" 这个动作
                    self.bus.read_operation_frame(self.motor_name)
                    
                time.sleep(0.02) # 50Hz 控制频率
                
            except Exception as e:
                # 忽略超时，因为这在 read_operation_frame 中很常见
                if "No response from the motor" not in str(e):
                    print(f"⚠️ 通信错误: {e}")
                time.sleep(0.5)

    def set_angle(self, angle_degrees: float):
        """设置目标角度（单位：度）"""
        # 限制范围，例如 +/- 2 圈
        angle_degrees = max(-720.0, min(720.0, angle_degrees))
        # target_position 是线程安全的 (原子操作)
        self.target_position = math.radians(angle_degrees)
        print(f" -> 目标设定: {angle_degrees:.1f}°")

    def set_kp(self, kp: float):
        """设置刚度"""
        if 0 <= kp <= 500:
            self.kp = kp
            print(f" -> 刚度(Kp)设定: {self.kp:.1f}")
        else:
            print("❌ Kp 范围必须在 0-500")

    def set_kd(self, kd: float):
        """设置阻尼"""
        if 0 <= kd <= 5:
            self.kd = kd
            print(f" -> 阻尼(Kd)设定: {self.kd:.1f}")
        else:
            print("❌ Kd 范围必须在 0-5")

    def stop_and_exit(self):
        print("\n🛑 正在停止...")
        self.running = False
        
        if self.control_thread:
            self.control_thread.join(timeout=0.5) # 等待线程退出
        
        if self.bus and self.connected:
            try:
                with self.lock:
                    # 回到零位
                    print("🏠 回到零位...")
                    self.bus.write_operation_frame(self.motor_name, 0.0, self.kp, self.kd, 0.0, 0.0)
                    time.sleep(1.0) # 等待电机移动
                    # 禁用
                    print("🚫 禁用电机...")
                    self.bus.disable(self.motor_name)
            except Exception as e:
                print(f"⚠️ 停止时出错: {e}")
            finally:
                self.bus.disconnect()
        
        print("👋 程序结束")
        sys.exit(0)

    def run_interactive(self):
        print("\n" + "="*40)
        print(f"🎮 MIT 位置控制台 (ID: {self.motor_id})")
        print("="*40)
        print("👉 直接输入数字 (单位: 度) 回车即可改变位置")
        print("👉 'kp <值>' (例如: kp 20) 来调节刚度 (消除抖动)")
        print("👉 'kd <值>' (例如: kd 0.8) 来调节阻尼 (消除抖动)")
        print("👉 '0' 或 'home' 回到零点")
        print("👉 'q' 退出")
        print(f"⚠️  当前 Kp={self.kp} | Kd={self.kd}")
        print("-" * 40)

        while True:
            try:
                cmd = input(f"[{math.degrees(self.target_position):.1f}°] >> ").strip().lower()
                
                if not cmd:
                    continue
                    
                if cmd in ['q', 'quit', 'exit']:
                    break
                
                if cmd in ['0', 'home']:
                    self.set_angle(0.0)
                    continue

                if cmd.startswith("kp "):
                    try:
                        new_kp = float(cmd.split()[1])
                        self.set_kp(new_kp)
                    except Exception:
                        print("❌ 无效 Kp. 示例: kp 20.0")
                    continue

                if cmd.startswith("kd "):
                    try:
                        new_kd = float(cmd.split()[1])
                        self.set_kd(new_kd)
                    except Exception:
                        print("❌ 无效 Kd. 示例: kd 0.5")
                    continue

                try:
                    angle = float(cmd)
                    self.set_angle(angle)
                except ValueError:
                    print("❌ 无效输入，请输入数字 (角度) 或 'kp', 'kd'")

            except KeyboardInterrupt:
                break
        
        self.stop_and_exit()

def main():
    if len(sys.argv) < 2:
        print("用法: python3 position_control_mit.py <motor_id>")
        sys.exit(1)
        
    motor_id = int(sys.argv[1])
    
    controller = PositionControllerMIT(motor_id)
    signal.signal(signal.SIGINT, controller._signal_handler)
    signal.signal(signal.SIGTERM, controller._signal_handler)
    
    if controller.connect():
        controller.run_interactive()

if __name__ == "__main__":
    main()

File: ref/RobStride_Control/cpp/src/position_control.cpp
/*
 * RobStride MIT 模式位置控制 (C++ 版本)
 * 模式: Mode 0 (MIT Mode)
 * 通信: 循环调用 write_operation_frame
 *
 * 编译:
 * g++ -o position_control_mit position_control_mit.cpp -lpthread -std=c++17
 *
 * 运行:
 * sudo ./position_control_mit <motor_id>
 * (需要 sudo 权限来访问 CAN 硬件)
 */

#include <iostream>
#include <string>
#include <cstring>
#include <thread>
#include <chrono>
#include <atomic>
#include <vector>
#include <csignal>
#include <cmath>

// Linux SocketCAN 头文件
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <linux/can.h>
#include <linux/can/raw.h>

// --- 全局原子变量，用于线程间安全通信 ---
std::atomic<bool> running(true);
std::atomic<double> target_position(0.0);
std::atomic<double> kp(100.0); // 高刚度
std::atomic<double> kd(2.0);  // 高阻尼

// --- 电机和协议定义 ---
const int MOTOR_ID_DEFAULT = 11; // 默认电机ID
const char* CAN_INTERFACE = "can0";
const int HOST_ID = 0xFF;

// 通信类型
const uint32_t COMM_OPERATION_CONTROL = 1;
const uint32_t COMM_ENABLE = 3;
const uint32_t COMM_WRITE_PARAMETER = 18;

// 参数 ID
const uint16_t PARAM_MODE = 0x7005;
const uint16_t PARAM_VELOCITY_LIMIT = 0x7017;
const uint16_t PARAM_TORQUE_LIMIT = 0x700B; // 注意: 根据 protocol.py 应该是 0x700B

// --- CAN 帧打包辅助函数 ---

// 将 float 复制到 uint8_t 数组中 (小端)
void pack_float_le(uint8_t* buf, float val) {
    memcpy(buf, &val, sizeof(float));
}

// 将 uint16_t 复制到 uint8_t 数组中 (小端)
void pack_u16_le(uint8_t* buf, uint16_t val) {
    memcpy(buf, &val, sizeof(uint16_t));
}

// 将 uint16_t 打包为大端字节序
void pack_u16_be(uint8_t* buf, uint16_t val) {
    buf[0] = (val >> 8) & 0xFF;
    buf[1] = val & 0xFF;
}

// --- 底层 CAN 函数 ---

/**
 * @brief 发送一个 CAN 帧
 */
bool send_frame(int s, uint32_t can_id, const uint8_t* data, uint8_t dlc) {
    struct can_frame frame;
    frame.can_id = can_id | CAN_EFF_FLAG; // 启用扩展帧 (29-bit)
    frame.can_dlc = dlc;
    if (data) {
        memcpy(frame.data, data, dlc);
    } else {
        memset(frame.data, 0, 8);
    }

    if (write(s, &frame, sizeof(struct can_frame)) != sizeof(struct can_frame)) {
        perror("write");
        return false;
    }
    return true;
}

/**
 * @brief 读取一个 CAN 帧 (带超时)
 */
bool read_frame(int s, struct can_frame* frame) {
    // 设置 100ms 超时
    struct timeval tv;
    tv.tv_sec = 0;
    tv.tv_usec = 100000; // 100ms
    fd_set rdfs;
    FD_ZERO(&rdfs);
    FD_SET(s, &rdfs);

    int ret = select(s + 1, &rdfs, NULL, NULL, &tv);
    if (ret == -1) {
        perror("select");
        return false;
    } else if (ret == 0) {
        // std::cerr << "read timeout" << std::endl;
        return false; // 超时
    }

    if (read(s, frame, sizeof(struct can_frame)) < 0) {
        perror("read");
        return false;
    }
    return true;
}

// --- RobStride 协议函数 ---

bool enable_motor(int s, int motor_id) {
    uint32_t ext_id = (COMM_ENABLE << 24) | (HOST_ID << 8) | motor_id;
    return send_frame(s, ext_id, nullptr, 0);
}

bool set_mode_raw(int s, int motor_id, int8_t mode) {
    uint32_t ext_id = (COMM_WRITE_PARAMETER << 24) | (HOST_ID << 8) | motor_id;
    uint8_t data[8] = {0};
    pack_u16_le(&data[0], PARAM_MODE); // param ID
    data[4] = (uint8_t)mode;           // value (int8)
    return send_frame(s, ext_id, data, 8);
}

bool write_limit(int s, int motor_id, uint16_t param_id, float limit) {
    uint32_t ext_id = (COMM_WRITE_PARAMETER << 24) | (HOST_ID << 8) | motor_id;
    uint8_t data[8] = {0};
    pack_u16_le(&data[0], param_id); // param ID
    pack_float_le(&data[4], limit);  // value (float)
    return send_frame(s, ext_id, data, 8);
}

bool write_operation_frame(int s, int motor_id, double pos, double kp_val, double kd_val) {
    // 1. 打包数据 (大端序!)
    // 这些 scaling value 应该从 table.py 导入，这里为了简化硬编码
    const double POS_SCALE = 4 * M_PI; // rs-03
    const double VEL_SCALE = 50.0;     // rs-03
    const double KP_SCALE = 5000.0;    // rs-03
    const double KD_SCALE = 100.0;     // rs-03
    const double TQ_SCALE = 60.0;      // rs-03

    // 裁剪和转换
    double pos_clamped = std::max(-POS_SCALE, std::min(POS_SCALE, pos));
    double kp_clamped = std::max(0.0, std::min(KP_SCALE, kp_val));
    double kd_clamped = std::max(0.0, std::min(KD_SCALE, kd_val));
    
    uint16_t pos_u16 = (uint16_t)(((pos_clamped / POS_SCALE) + 1.0) * 0x7FFF);
    uint16_t vel_u16 = 0x7FFF; // 0 velocity
    uint16_t kp_u16 = (uint16_t)((kp_clamped / KP_SCALE) * 0xFFFF);
    uint16_t kd_u16 = (uint16_t)((kd_clamped / KD_SCALE) * 0xFFFF);
    uint16_t torque_u16 = 0x7FFF; // 0 torque_ff

    uint8_t data[8];
    pack_u16_be(&data[0], pos_u16);
    pack_u16_be(&data[2], vel_u16);
    pack_u16_be(&data[4], kp_u16);
    pack_u16_be(&data[6], kd_u16);
    
    // 2. 构建 CAN ID
    uint32_t ext_id = (COMM_OPERATION_CONTROL << 24) | (torque_u16 << 8) | motor_id;
    
    // 3. 发送
    return send_frame(s, ext_id, data, 8);
}

bool read_operation_frame(int s) {
    struct can_frame frame;
    if (read_frame(s, &frame)) {
        if (!frame.can_id & CAN_EFF_FLAG) return false;
        
        uint32_t comm_type = (frame.can_id >> 24) & 0x1F;
        if (comm_type == 2) { // 状态包
            return true;
        }
    }
    return false;
}

/**
 * @brief 控制循环线程
 */
void control_loop(int s, int motor_id) {
    std::cout << "🔄 控制循环已启动 (Mode 0 @ 50Hz)" << std::endl;
    
    while (running) {
        auto start = std::chrono::steady_clock::now();
        
        // 1. 发送 MIT 帧 (只发)
        write_operation_frame(s, motor_id, target_position.load(), kp.load(), kd.load());
        
        // 2. 读取状态帧 (只收)，清空缓冲区
        while(read_operation_frame(s)); // 循环读取，直到缓冲区为空或超时
        
        auto end = std::chrono::steady_clock::now();
        auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        
        // 固定的 50Hz 循环
        std::this_thread::sleep_for(std::chrono::microseconds(20000) - elapsed);
    }
    std::cout << "⏹️ 控制线程停止" << std::endl;
}

/**
 * @brief 初始化 SocketCAN
 */
int init_can(const char* ifname) {
    int s;
    struct sockaddr_can addr;
    struct ifreq ifr;

    if ((s = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
        perror("socket");
        return -1;
    }

    strcpy(ifr.ifr_name, ifname);
    if (ioctl(s, SIOCGIFINDEX, &ifr) < 0) {
        perror("ioctl");
        close(s);
        return -1;
    }

    addr.can_family = AF_CAN;
    addr.can_ifindex = ifr.ifr_ifindex;

    if (bind(s, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("bind");
        close(s);
        return -1;
    }

    return s;
}

void signal_handler(int signum) {
    std::cout << "\n🛑 捕获到退出信号..." << std::endl;
    running = false;
}

int main(int argc, char* argv[]) {
    int motor_id = MOTOR_ID_DEFAULT;
    if (argc > 1) {
        motor_id = std::atoi(argv[1]);
    }
    
    std::cout << "🎯 MIT 位置控制台 (ID: " << motor_id << ")" << std::endl;
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    int s = init_can(CAN_INTERFACE);
    if (s < 0) {
        std::cerr << "❌ 无法打开 CAN 接口 " << CAN_INTERFACE << std::endl;
        return 1;
    }
    std::cout << "📡 CAN 总线 " << CAN_INTERFACE << " 连接成功" << std::endl;

    // --- 初始化电机 ---
    std::cout << "⚡ 激活电机 ID: " << motor_id << " ..." << std::endl;
    enable_motor(s, motor_id);
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    
    std::cout << "⚙️ 切换到 MIT 模式 (Mode 0)..." << std::endl;
    set_mode_raw(s, motor_id, 0);
    
    std::cout << "⚙️ 设置内部限制..." << std::endl;
    write_limit(s, motor_id, PARAM_VELOCITY_LIMIT, 20.0);
    write_limit(s, motor_id, PARAM_TORQUE_LIMIT, 20.0);
    
    std::cout << "🏠 设置初始目标为 0.0 ..." << std::endl;
    write_operation_frame(s, motor_id, 0.0, kp.load(), kd.load());
    std::cout << "✅ 初始化完成！" << std::endl;

    // 启动控制线程
    std::thread t(control_loop, s, motor_id);

    // --- 交互式主循环 ---
    std::cout << "\n" << "========================================" << std::endl;
    std::cout << "👉 输入数字 (度) 回车即可改变位置" << std::endl;
    std::cout << "👉 'kp <值>' (例如: kp 100) 来调节刚度" << std::endl;
    std::cout << "👉 'kd <值>' (例如: kd 2.0) 来调节阻尼 (防抖)" << std::endl;
    std::cout << "👉 '0' 或 'home' 回到零点" << std::endl;
    std::cout << "👉 'q' 退出" << std::endl;
    std::cout << "⚠️  当前 Kp=" << kp.load() << " | Kd=" << kd.load() << std::endl;
    std::cout << "----------------------------------------" << std::endl;

    std::string line;
    while (running) {
        double current_pos_deg = target_position.load() * 180.0 / M_PI;
        printf("[%.1f°] >> ", current_pos_deg);
        std::flush(std::cout);

        if (!std::getline(std::cin, line)) {
            break;
        }

        if (line == "q" || line == "quit" || line == "exit") {
            running = false;
        } else if (line == "0" || line == "home") {
            target_position = 0.0;
            std::cout << " -> 目标设定: 0.0°" << std::endl;
        } else if (line.rfind("kp ", 0) == 0) {
            try {
                kp = std::stod(line.substr(3));
                std::cout << " -> 刚度(Kp)设定: " << kp.load() << std::endl;
            } catch (...) {
                std::cout << "❌ 无效 Kp. 示例: kp 100.0" << std::endl;
            }
        } else if (line.rfind("kd ", 0) == 0) {
            try {
                kd = std::stod(line.substr(3));
                std::cout << " -> 阻尼(Kd)设定: " << kd.load() << std::endl;
            } catch (...) {
                std::cout << "❌ 无效 Kd. 示例: kd 2.0" << std::endl;
            }
        } else {
            try {
                double angle_deg = std::stod(line);
                angle_deg = std::max(-720.0, std::min(720.0, angle_deg));
                target_position = angle_deg * M_PI / 180.0;
                std::cout << " -> 目标设定: " << angle_deg << "°" << std::endl;
            } catch (...) {
                std::cout << "❌ 无效输入" << std::endl;
            }
        }
    }

    // 清理
    running = false;
    t.join();
    
    std::cout << "🏠 回到零位..." << std::endl;
    write_operation_frame(s, motor_id, 0.0, kp.load(), kd.load());
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    
    std::cout << "🚫 禁用电机..." << std::endl;
    // C++ SDK 中没有 disable, 我们手动发送一个 kp=0, kd=0 的帧
    write_operation_frame(s, motor_id, 0.0, 0.0, 0.0);
    
    close(s);
    std::cout << "👋 程序结束" << std::endl;
    
    return 0;
}